src/commands/buffer.rs

```diff
@@ -468,6 +469,127 @@ pub fn outdent_line(app: &mut Application) -> Result {
+            m.insert(Scope::new("embedding.php").unwrap(), "// ");
+            m.insert(Scope::new("source.c").unwrap(), "// ");
+            m.insert(Scope::new("source.c++").unwrap(), "// ");
+            m.insert(Scope::new("source.cmake").unwrap(), "# ");
+            m.insert(Scope::new("source.go").unwrap(), "// ");
+            m.insert(Scope::new("source.js").unwrap(), "// ");
+            m.insert(Scope::new("source.lua").unwrap(), "-- ");
+            m.insert(Scope::new("source.makefile").unwrap(), "# ");
+            m.insert(Scope::new("source.python").unwrap(), "# ");
+            m.insert(Scope::new("source.rust").unwrap(), "// ");
+            m.insert(Scope::new("source.shell.bash").unwrap(), "# ");
+            m.insert(Scope::new("source.swift").unwrap(), "// ");
+            m.insert(Scope::new("source.toml").unwrap(), "# ");
+            m.insert(Scope::new("source.yaml").unwrap(), "# ");
+            m.insert(Scope::new("text.tex").unwrap(), "% ");
+            m.insert(Scope::new("text.tex.latex").unwrap(), "% ");
+            m
+        };
+    }
+
+    let buffer = app.workspace.current_buffer().ok_or(BUFFER_MISSING)?;
+    let original_cursor = *buffer.cursor.clone();
+
+    let comment_content = {
+        let scope = buffer.syntax_definition.as_ref()
+            .map_or(Scope::new("none").unwrap(), |def| def.scope);
+
+        COMMENT_MAP.get(&scope).ok_or_else(|| {
+            format!("No comment syntax definition found for scope {}", scope)
+        })?
+    };
+
+    // Get the range of lines we'll comment based on
+    // either the current selection or cursor line.
+    let lines = match app.mode {
+        Mode::SelectLine(ref mode) => {
+            if mode.anchor >= buffer.cursor.line {
+                buffer.cursor.line..mode.anchor + 1
+            } else {
+                mode.anchor..buffer.cursor.line + 1
+            }
+        }
+        _ => buffer.cursor.line..buffer.cursor.line + 1,
+    };
+
+    let buffer_range = Range::new(
+        Position { line: lines.start, offset: 0 },
+        Position { line: lines.end, offset: 0 }
+    );
+
+    let buffer_range_content = buffer.read(&buffer_range).ok_or(CURRENT_LINE_MISSING)?;
+
+    // Split the buffer content on newlines and filter out any zero-length lines.
+    let lines: Vec<(usize, &str)> = lines
+        .zip(buffer_range_content.split("\n"))
+        .filter(|(_, line)| line.trim().len() > 0)
+        .collect();
+
+    // We look at all lines to see if they start with `comment_content` or not.
+    // If even a single line does not, we need to comment all lines out,
+    // otherwise remove `comment_content` on the start of each line.
+    let (toggle, offset) = lines.iter()
+        .map(|(_, line)| {
+            let content = line.trim_start();
+            (content.starts_with(comment_content), line.len() - content.len())
+        })
+        .fold((true, usize::MAX), |result, (toggle, offset)| {
```

Comment:
They all will be indented at the lowest common level. To demonstrate what I mean, I made a [short gif](https://giphy.com/gifs/ZiOh0JIcaDwtPjX5cl/html5). (Could have put that in the PR from the beginning, I guess.)
That's how I had the neovim plugin configured. I personally like that more than indenting each line at the start of its content, but that would be doable too. (Maybe even as a configuration item?)

The combination of reducing to the toggle switch and offset as tuple is only to avoid iterating all the lines twice. I could split that for clarity, but at cost.

Reply:
Oh, of course; I don't know what I was thinking, that's exactly the behaviour we want. It seemed odd when reading the source code first, before having really thought about the behaviour.